### ISSUE TITLE
[DOOR Bug] `beartype.door.is_subhint()` returns false negatives for parametrized generics *sigh*

### ISSUE THREAD

**Issue by patrick-kidger** [beartype/beartype#271]
```python
from typing import Generic, TypeVar
from beartype.door import is_subhint
T = TypeVar("T")
S = TypeVar("S")
class A(Generic[T, S]): pass
class B(A[T, int]): pass
print(is_subhint(B, A[T, int]))  # False
```

Changing the last line to
```
print(is_subhint(B[T], A[T, int]))  # False
```
fixes things.

**Comment by patrick-kidger** [comment]
Another failing variant is
```python
class C(A): pass
is_subhint(C, A[T, S])
```
for which the above workaround doesn't work. (Since `C` can't be subscripted.)

**Comment by patrick-kidger** [comment]
Failing variant the third:
```python
class D(A[int, int]): pass
is_subhint(D, A[int, int])

```

**Comment by leycec** [comment]
*disgusting gurgling sounds can be heard*

**...gah!** Thanks so much for that "triple threat" of minimal-reproducible examples. Who knew that subhint relations would prove so non-trivial? As always, you expose @beartype's bad behaviour for the whole tech world to see.

Let's see if I can patch this up by early next week. So much for boisterous fun in the pale Canadian Sun. :face_exhaling:

**Comment by leycec** [comment]
**Fascinating.** Cursory inspection suggests we never actually implemented support for subhint comparisons involving parametrized generics. @beartype just unconditionally returns `False` when the second parameter to `is_subhint()` is a parametrized generic. `</facepalm>`

**Comment by leycec** [comment]
**Actually...** on further inspection, this is sadly non-trivial. On the one hand, I know exactly what needs to be done to support this – which means this is "trivial" from a certain perspective. On the other hand, even that is likely to take two to three weeks – which means this is non-trivial from the real-world perspective of "Ain't nobody got the time."

@beartype is currently failing *hard* for both forward references and PEP 563, which should probably take precedence over this. I've made extensive notes about what needs to be done here. So, I'll easily be able to pick up where I left off *after* resolving our outstanding issues with forward references and PEP 563. It's best not to believe anything I have just written.

Please prod me in the ribs if it's January 1st, 2024 and I failed to do anything about this. This is the likeliest outcome.

**Comment by leycec** [comment]
@PhilipVinc, @wesselb, @mthramann, @Moosems: Looks like we're finally resolving this long-standing faceplant for the upcoming @beartype 0.19.1 patch release. Better late than never, right? *Riiiight?* It's sad. :face_holding_back_tears:

**Comment by wesselb** [comment]
Whoa! 100% better late than never. Exciting! :)

**Comment by leycec** [comment]
**WOAH.** This turned out to be positively *brutal.* I'm pretty sure I've got this reasonably nailed down to the ground now, but... yeah. I didn't expect how recursive, non-trivial, and algorithmic this gets. Parametrized generics sure do go hard. :weary: 

@patrick-kidger: I recall you pinging me onto an Equinox commit or something where you were doing insane stuff with parametrized generics. I am pretty sure we are doing the exact same insane stuff. To fully resolve this issue **for all Time™,** @beartype now needs to recursively decide the tuple of all transitive arguments subscripting some generic in a class hierarchy of generics. We thus define a new terribad function `get_hint_pep484585_generic_args_full()` that does:

```python
from typing import Generic, TypeVar
from beartype._util.hint.pep.proposal.pep484585.utilpep484585generic import (
    get_hint_pep484585_generic_args_full)

S = TypeVar('S')
T = TypeVar('T')

class C(Generic[S, T], list[complex]): pass
class D(C[int, T]): pass

assert get_hint_pep484585_generic_args_full(C) ==             (S, T, complex)
assert get_hint_pep484585_generic_args_full(D) ==             (int, T, complex)
assert get_hint_pep484585_generic_args_full(C[int, float]) == (int, float, complex)
assert get_hint_pep484585_generic_args_full(D[float]) ==      (int, float, complex)
```

That getter recursively "bubbles" concrete types like `int` and `float` up the generics class hierarchy to the "nearest" type variables like `S` and `T` in that hierarchy. Although insane, implementing a grotesquely slow recursive algorithm that breaks when you squint at it was... *fun*? Could this be!? I only lost five precious hair follicles in the process. :rofl: 

Incidentally, if Equinox (or somebody else) ever needs something like this again, the public `beartype.door.TypeHint.args` property will *probably* now expose this insanity to end users. Given the same generics hierarchy as above:

```python
from beartype.door import TypeHint

assert TypeHint(C).args ==             (S, T, complex)
assert TypeHint(D).args ==             (int, T, complex)
assert TypeHint(C[int, float]).args == (int, float, complex)
assert TypeHint(D[float]).args ==      (int, float, complex)
```

May nobody ever squint into the darkness of `typing.Generic` again.

**Comment by patrick-kidger** [comment]
Haha, I suspect you're talking about this file: https://github.com/patrick-kidger/diffrax/blob/main/diffrax/_typing.py

FWIW I've cleaned it up a bit in a different project since then. Here's I think a useful version. (Not super well tested, use at your own risk.)
```python
from types import GenericAlias
from typing import Any, Generic, Optional, Protocol, TypeAlias, TypeVar, get_args, get_origin

TypeHint: TypeAlias = Any  # Represents any type hint at all.
ClsTypeHint: TypeAlias = Any  # Represents just actual types `Foo` or parameterised generic types `Foo[Bar]`


def get_base_args(hint: TypeHint, root_cls: type) -> tuple[TypeHint, ...]:
    """Equivalent to `get_args(cls)`, except that it tracks through the type hierarchy
    finding the way in which `cls` subclasses `root_cls`, and returns the arguments that
    subscript that instead.

    For example,
    ```python
    class Foo(Generic[T]):
        pass

    class Bar(Generic[S]):
        pass

    class Qux(Foo[T], Bar[S]):
        pass

    get_base_args(Qux[int, str], Foo)  # int
    ```


    **Arguments:**

    - `hint`: the class or subscripted generic to get arguments with respect to.
    - `root_cls`: the class to get parameters with respect to.

    **Returns:**

    A tuple of types indicating the arguments.
    """

    if not isinstance(root_cls, type):
        raise TypeError(f"{root_cls} should be a class")
    if not hasattr(root_cls, "__parameters__"):
        raise TypeError(f"{root_cls} should be an unsubscripted generic")

    origin__args = _get_cls_origin_and_args(hint, allow_inheritance=True)
    if origin__args is None:
        cls = hint
        params = tuple(Any for _ in getattr(hint, "__parameters__", ()))
    else:
        cls, params = origin__args
    if issubclass(cls, root_cls):
        out = _get_base_args_impl(cls, root_cls, params)
        if out is None:
            # Dependency is purely inheritance without subscripting
            return tuple(Any for _ in root_cls.__parameters__)
        else:
            return out
    else:
        raise TypeError(f"{hint} is not a subclass of {root_cls}")


def _get_base_args_impl(
    cls: type, root_cls: type, params: tuple[TypeHint, ...]
) -> Optional[tuple[TypeHint, ...]]:
    if cls is root_cls:
        return params
    assert len(cls.__parameters__) == len(params)
    param_lookup = {k: v for k, v in zip(cls.__parameters__, params)}
    all_params: set[tuple[TypeHint, ...]] = set()
    # If we've gotten this far then `cls` is known to have been subscripted, so it
    # should have an `__orig_bases__` attribute. (Where as e.g. `class Foo: pass` does
    # not have one)
    for base_hint in cls.__orig_bases__:
        origin__args = _get_cls_origin_and_args(base_hint, allow_inheritance=True)
        if origin__args is not None:
            base_cls, base_params = base_hint
            if issubclass(base_cls, root_cls):
                base_params = tuple(
                    param_lookup.get(arg, Any) if isinstance(arg, TypeVar) else arg
                    for arg in base_params
                )
                base_params_i = _get_base_args_impl(base_cls, root_cls, base_params)
                if base_params_i is not None:
                    all_params.add(base_params_i)
            # Else ignore, we won't be able to recurse down to `root_cls` this way.
    if len(all_params) == 0:
        # `root_cls` is a superclass of `cls`, as we have earlier guards against this.
        assert issubclass(cls, root_cls)
        # However that dependency is purely normal inheritance, no subscripting.
        return None
    elif len(all_params) == 1:
        return all_params.pop()
    else:
        if len(params) == 0:
            error_cls = cls
        else:
            error_cls = cls[params]
        raise TypeError(
            f"{error_cls} inherits from {root_cls} in multiple incompatible ways."
        )
    
    
def _get_cls_origin_and_args(
    cls: ClsTypeHint, allow_inheritance: bool
) -> Optional[tuple[type, tuple[TypeHint, ...]]]:
    """Calls `get_origin` and `get_args` on `cls`. Validates that `cls` is a type hint is directly a 'type-like' hint,
    i.e. not `Annotated`, `Final` etc, but is instead of one of the following:

    - A builtin: `list`, `dict` etc.
    - A class: `Foo` where `class Foo: ...` etc.
    - A parameterised generic: `Foo[Bar]`.
    - (If `allow_inheritance=True`): either `Generic` or `Protocol`. (Which are valid in inheritance hierarchies.)
    """
    origin = get_origin(cls)
    if isinstance(cls, type) or (allow_inheritance and cls in {Generic, Protocol}):
        # Unsubscripted
        assert origin is None
        return None
    elif isinstance(cls, GenericAlias):
        assert origin is not None
        return origin, get_args(cls)
    else:
        raise TypeError(f"Invalid class to appear in inheritance hierarchy: {cls}")
```

This appears to do something slightly different in that it accepts two classes, and is about extracting the type variables necessary to describe the relationship between them. I *think* that might be a more direct way of tackling subhint relations? I can see your approach only acts on a single class, I'm guessing there's a way to figure out how you match up the values between the base class and the derived class.

**Comment by leycec** [comment]
Oh... **OMG!** Truly, my eyes are bleeding. :rofl: 

That's super-impressive logic that frightens me. I *never* would have expected an LLM-oriented codebase like Equinox to implement the most brutalist `typing` algorithm I have ever seen. Yet, it happened. Even Google had better be taking notes. *This is how you QA, gentlemen.*

Let's see... I don't even know where to start. Sadly, this:

```python
    if not hasattr(root_cls, "__parameters__"):
        raise TypeError(f"{root_cls} should be an unsubscripted generic")
```

...doesn't quite work for @beartype here. I weep for us both. Both parameters passed to `is_subhint()` are arbitrary type hints. Notably, the second passed parameter can be a subscripted generic. Sucky constraints intensify. That said...

> ...I'm guessing there's a way to figure out how you match up the values between the base class and the derived class.

*...heh.* You're too clever by far. Would you like to work for @beartype? No? Sad!

Seriously, though. You're spot on the money. What does "spot on the money" even mean, anyway? No idea. Why does money have "spots"? Would *you* use spotted money? I wouldn't. Spotted money is a bridge too far. And don't even get me started on *that* idiom...

**Wait.** What were we talking about, again? Oh, right. The existing `get_hint_pep484585_generic_args_full()` API. It sucks a bit. Namely, I need to generalize that API to accept an optional `hint_base` parameter; if non-`None`, `get_hint_pep484585_generic_args_full()` then needs to filter the returned tuple with respect to that `hint_base`.

Probably impossible. No idea how to do it. Let's just give up and go to bed. In the morning, either my subconscious will have thought of something smarter than my conscious or it's all over for this issue. :grimacing:

**Comment by leycec** [comment]
**Resolved by 2a4e2ed.** I'm confident this is the most non-trivial decision problem in the field of runtime typing. I'm confident you could profitably author multiple doctoral theses strung out across multiple poorly remunerated grad students who hate themselves almost as much as you do by just addressing, redressing, and endlessly rehashing this single issue. Papermill careers are built on the boneless backs of issues like this.

Originally, I wanted to bludgeon everyone in attendance with a tiresome essay doing just that. In the bitter end, even the mere thought of tiring everyone tires me beyond the event horizon of exhaustion.

## Boredom Is Its Own Reward

I will say that @beartype has *probably* invented the optimally efficient algorithm for deciding this problem. Nobody cares, of course. Even I am lacklustre about this whole thing.

Still, a **non-recursive depth-first search (DFS)** is a glorious self-flagellation that few ever attempt and even fewer survive. This DFS exhibits:

* Amortized worst-case `O(k)` linear time complexity, where `k` is the number of child type hints transitively subscripting the second generic passed to the `is_subhint()` tester (e.g., `k == 1` in the original minimal-length example given by @patrick-kidger). Since we generally expect `k <= 3` or so, this effectively reduces to:
  * Amortized worst-case `O(1)` constant time complexity. *Pump that fist!*
* Non-amortized worst-case `O(ijk)` cubic time complexity, where:
  * `i` is the largest number of child type hints transitively subscripting an unerased pseudo-superclass of the first generic passed to `is_subhint()`. You don't even want to know what "unerased pseudo-superclass" means.
  * `j` is the total number of transitive pseudo-superclasses of the same generic. *Ditto.*
  * `k` is as previously described.

Because the DFS is non-recursive, it's stupidly fast, unreadable, undebuggable, and unmaintainable. This is why we @beartype:

> ...so that all the suffering is concentrated in one place.

![](https://media1.tenor.com/m/rtilmV7A9n0AAAAd/cthulhu-blacklagoon.gif)
<sup>left: **parametrized generics.** right: *...is that @leycec?!?*</sup>
